### PR TITLE
PS-7310: Merge fb-prod8-202007 to fb-prod8-202009 (8.0)

### PR DESCRIPTION
**Pull request by inikep**
1. Add new MyRocks variable `rocksdb_max_compaction_history`.

2. Deprecate MyRocks variables `strict_collation_check` and `strict_collation_exceptions`. All collations
  are now supported for `CHAR` and `VARCHAR` indexed columns.

3. Add new MyRocks Information Schema Tables `ROCKSDB_ACTIVE_COMPACTION_STATS` and `ROCKSDB_COMPACTION_HISTORY`.
https://www.percona.com/doc/percona-server/LATEST/myrocks/information-schema-tables.html

4. Update rocksdb submodule to [12b78e40bdb Track WAL in MANIFEST: add option track_and_verify_wals_in_manifest ](https://github.com/facebook/rocksdb/commit/12b78e40bdb)

5. Cherry-pick the following commits from `fb-prod8-202007`:
- [01212895d09 remove unused functions ha_rocksdb::remove_rows](https://github.com/facebook/mysql-5.6/commit/01212895d09)
- [8027f8906b0 Fix leak in Rdb_ha_data](https://github.com/facebook/mysql-5.6/commit/8027f8906b0)
- [168a481227c Extend support to add more active compaction jobs details from RocksDB in MyRocks.](https://github.com/facebook/mysql-5.6/commit/168a481227c)
- [8221b82141f Added more descriptive error when parsing cf options in RocksDB](https://github.com/facebook/mysql-5.6/commit/8221b82141f)
- [9e4c27ba937 update rocksdb_create_temporary_checkpoint testcase](https://github.com/facebook/mysql-5.6/commit/9e4c27ba937)
- [de1de74ba1c Stabilize rocksdb.drop_cf_during_show_global_info in valgrind](https://github.com/facebook/mysql-5.6/commit/de1de74ba1c)
- [a4e6e6b0381 Fix cleanup on rocksdb.drop_cf_before_show_deadlock_info test](https://github.com/facebook/mysql-5.6/commit/a4e6e6b0381)
- [f638f6e85b2 Version-independent key packing](https://github.com/facebook/mysql-5.6/commit/f638f6e85b2)
- [1b31d40a854 Make BIT, ENUM, SET covering for MyRocks](https://github.com/facebook/mysql-5.6/commit/1b31d40a854)
- [1046d4f7074 Block Alter table with string indexed column to non-binary collation.](https://github.com/facebook/mysql-5.6/commit/1046d4f7074)
- [5feccbd9c65 FB8-246 main.implicit_commit fails in debug compilation (#1117)](https://github.com/facebook/mysql-5.6/commit/5feccbd9c65)
- [9896a5d6087 add `INFORMATION_SCHEMA.ROCKSDB_COMPACTION_HISTORY` table](https://github.com/facebook/mysql-5.6/commit/9896a5d6087)
- [d26314fc8aa Optimize pack_with_varchar_space_pad handling for empty data](https://github.com/facebook/mysql-5.6/commit/d26314fc8aa)
- [e74eb32cef9 Support covering prefixed key for blob/text fields](https://github.com/facebook/mysql-5.6/commit/e74eb32cef9)
- [ffbdbbec15d Added the optimizations in bitmap used for covering Index.](https://github.com/facebook/mysql-5.6/commit/ffbdbbec15d)
- [c38941b14df Save 2 bytes length from unpack_info value for secondary index.](https://github.com/facebook/mysql-5.6/commit/c38941b14df)
- [74605f6fde5 Better MyRocks init failure cleanup](https://github.com/facebook/mysql-5.6/commit/74605f6fde5)
- [6dba98bac56 Use rocksdb::Status::IsPathNotFound instead of errno](https://github.com/facebook/mysql-5.6/commit/6dba98bac56)

6. Cherry-pick the following commits from `fb-prod8-202008`:
- [2657ae17be55d2a358738977ac535baa70ad4fb5 Better cleanup in Myrocks initialization failure](https://github.com/facebook/mysql-5.6/commit/2657ae17be55d2a358738977ac535baa70ad4fb5)
- [9760980a8c1280e464b19de6951e801fa0be17d1 Fix bypass bug in unpack info](https://github.com/facebook/mysql-5.6/commit/9760980a8c1280e464b19de6951e801fa0be17d1)
- [3df831ac4ec65c2766fbfb3faf1519f4c783dae3 PS-6074: NO_PAD collations should not use the space padded encoding](https://github.com/facebook/mysql-5.6/commit/3df831ac4ec65c2766fbfb3faf1519f4c783dae3)
- [c78110e2266bcef33e160c963ea51a6dd2424364 PS-7027: MyRocks: Optimized support for utf8mb4_bin and other binary collations](https://github.com/facebook/mysql-5.6/commit/c78110e2266bcef33e160c963ea51a6dd2424364)
- [41d164795d6b8c1353f186520d45ff5ff5956d63 PS-7027: MyRocks: Add SK index-only scans for all remaining collations](https://github.com/facebook/mysql-5.6/commit/41d164795d6b8c1353f186520d45ff5ff5956d63)
- [49c633e09526a29f1bdc5fb26a493bd83170d3ed Add rocksdb_wsenv_path for initial WSEnvironment support](https://github.com/facebook/mysql-5.6/commit/49c633e09526a29f1bdc5fb26a493bd83170d3ed)
- [17f87c30ab8e9bfed448883e186be3ab6e0df3a6 PS-7027: Fix EXPLAIN SELECT for RocksDB SKs](https://github.com/facebook/mysql-5.6/commit/17f87c30ab8e9bfed448883e186be3ab6e0df3a6)
- [5f533bbdb4d934fb00f15e64a020f8cf4d31d353 convert db and table name to my_charset_filename during MyRocks DD validate](https://github.com/facebook/mysql-5.6/commit/5f533bbdb4d934fb00f15e64a020f8cf4d31d353)
- [0e64b53ef1dc9f53bfed02c81c0becdcfd902d78 Remove 'all fields in sk are covered read' assert](https://github.com/facebook/mysql-5.6/commit/0e64b53ef1dc9f53bfed02c81c0becdcfd902d78)
- [047268f9c0b8e6270534193b48693533c0e9a393 Use Rocksdb 6.12.1](https://github.com/facebook/mysql-5.6/commit/047268f9c0b8e6270534193b48693533c0e9a393)

7. Cherry-pick the following commits from `fb-prod8-202009`:
- [196e1b8ce2c515244d26a83fcbe48d503b30aad3 Making rocksdb_max_bottom_pri_background_compactions dynamic](https://github.com/facebook/mysql-5.6/commit/196e1b8ce2c515244d26a83fcbe48d503b30aad3)
- [2c0edff70c20284f367d953c7f5372077a6cce0f Fix direct_io check](https://github.com/facebook/mysql-5.6/commit/2c0edff70c20284f367d953c7f5372077a6cce0f)
- [6cc1ced51606d5d0da5db2b078765c93c188b040 Update to latest rocksdb and update MyRocks to fix build break](https://github.com/facebook/mysql-5.6/commit/6cc1ced51606d5d0da5db2b078765c93c188b040)

8. Skip the following commits that break backward compatibility:
- [7ab0fa58f41 Change to remove lagecy VARBINARY format](https://github.com/facebook/mysql-5.6/commit/7ab0fa58f41)
- [bbf6a97e8af Remove handling code for legacy format](https://github.com/facebook/mysql-5.6/commit/bbf6a97e8af)

### CODE REVIEW

**Comment by inikep** [comment]
Tested at https://ps80.cd.percona.com/job/percona-server-8.0-param/845/

File: storage/rocksdb/rdb_i_s.cc [hunk 1]
```diff
@@ -2004,6 +2076,23 @@ struct st_mysql_plugin rdb_i_s_compact_stats = {
     0,       /* flags */
 };
 
+struct st_mysql_plugin rdb_i_s_active_compact_stats = {
```
Comment (george-lorch) [inline storage/rocksdb/rdb_i_s.cc:2220]: We are going to need a documentation ticked to document this here https://www.percona.com/doc/percona-server/LATEST/myrocks/information-schema-tables.html

File: storage/rocksdb/rdb_i_s.cc [hunk 2]
```diff
@@ -2093,6 +2228,23 @@ struct st_mysql_plugin rdb_i_s_active_compact_stats = {
     0,       /* flags */
 };
 
+struct st_mysql_plugin rdb_i_s_compact_history = {
```
Comment (george-lorch) [inline storage/rocksdb/rdb_i_s.cc:2237]: We are going to need a doc ticket to be sure this gets documented here https://www.percona.com/doc/percona-server/LATEST/myrocks/information-schema-tables.html

File: storage/rocksdb/ha_rocksdb.cc [hunk 3]
```diff
@@ -1387,6 +1392,10 @@ static MYSQL_SYSVAR_ULONG(
     nullptr, nullptr, rocksdb_persistent_cache_size_mb,
     /* min */ 0L, /* max */ ULONG_MAX, 0);
 
+static MYSQL_SYSVAR_STR(wsenv_path, rocksdb_wsenv_path,
```
Comment (george-lorch) [inline storage/rocksdb/ha_rocksdb.cc:1395]: I'm not sure if we are building with this enabled or not, but if o we will need documentation ticket for it.

File: storage/rocksdb/ha_rocksdb.cc [hunk 4]
```diff
@@ -2126,6 +2150,15 @@ static MYSQL_SYSVAR_STR(
...
 
+static MYSQL_SYSVAR_ULONGLONG(
+    max_compaction_history, rocksdb_max_compaction_history, PLUGIN_VAR_OPCMDARG,
+    "Track history for at most this many completed compactions. "
+    "The history is in the INFORMATION_SCHEMA.ROCKSDB_COMPACTION_HISTORY "
+    "table.",
+    nullptr, rocksdb_max_compaction_history_update,
+    RDB_DEFAULT_MAX_COMPACTION_HISTORY, 0ULL /* min */, UINT64_MAX /* max */,
+    0 /* blk */);
+
```
Comment (george-lorch) [inline storage/rocksdb/ha_rocksdb.cc:2161]: Documentation ticket

**Review by george-lorch — COMMENTED** [review]
We will also need to create a doc ticket to update the limitations page here https://www.percona.com/doc/percona-server/LATEST/myrocks/limitations.html w.r.t. collations and support.

**Review by george-lorch — APPROVED** [review]

File: storage/rocksdb/rdb_i_s.cc (hunk 1 above)
Comment (inikep) [reply storage/rocksdb/rdb_i_s.cc:2220]: Already created at https://jira.percona.com/browse/PS-7442

File: storage/rocksdb/rdb_i_s.cc (hunk 2 above)
Comment (inikep) [reply storage/rocksdb/rdb_i_s.cc:2237]: Already created at https://jira.percona.com/browse/PS-7442

File: storage/rocksdb/ha_rocksdb.cc (hunk 4 above)
Comment (inikep) [reply storage/rocksdb/ha_rocksdb.cc:2161]: Already created at https://jira.percona.com/browse/PS-7441

File: storage/rocksdb/ha_rocksdb.cc (hunk 3 above)
Comment (inikep) [reply storage/rocksdb/ha_rocksdb.cc:1395]: We will not use it. I kept it just to minimize a diff with upstream.

**Pull request merged**